### PR TITLE
Release v1.5.1

### LINKED ISSUE #7992: Publish next Zebra release: 1.5.1
# Prepare for the Release

These release steps can be done a week before the release, in separate PRs.
They can be skipped for urgent releases.

## State Full Sync Test

To check consensus correctness, we want to test that the state format is valid after a full sync. (Format upgrades are tested in CI on each PR.)

- [ ] Make sure there has been [at least one successful full sync test](https://github.com/ZcashFoundation/zebra/actions/workflows/ci-integration-tests-gcp.yml?query=event%3Aschedule) since the last state change, or
- [ ] Start a manual workflow run with a Zebra and `lightwalletd` full sync.

State format changes can be made in `zebra-state` or `zebra-chain`. The state format can be changed by data that is sent to the state, data created within the state using `zebra-chain`, or serialization formats in `zebra-state` or `zebra-chain`. 

After the test has been started, or if it has finished already:
- [ ] Ask for a state code freeze in Slack. The freeze lasts until the release has been published.

## Checkpoints

For performance and security, we want to update the Zebra checkpoints in every release.
- [ ] You can copy the latest checkpoints from CI by following [the zebra-checkpoints README](https://github.com/ZcashFoundation/zebra/blob/main/zebra-utils/README.md#zebra-checkpoints).

## Missed Dependency Updates

Sometimes `dependabot` misses some dependency updates, or we accidentally turned them off.

This step can be skipped if there is a large pending dependency upgrade. (For example, shared ECC crates.)

Here's how we make sure we got everything:
- [ ] Run `cargo update` on the latest `main` branch, and keep the output
- [ ] If needed, [add duplicate dependency exceptions to deny.toml](https://github.com/ZcashFoundation/zebra/blob/main/book/src/dev/continuous-integration.md#fixing-duplicate-dependencies-in-check-denytoml-bans)
- [ ] If needed, remove resolved duplicate dependencies from `deny.toml`
- [ ] Open a separate PR with the changes
- [ ] Add the output of `cargo update` to that PR as a comment

# Prepare and Publish the Release

Follow the steps in the [release checklist](https://github.com/ZcashFoundation/zebra/blob/main/.github/PULL_REQUEST_TEMPLATE/release-checklist.md) to prepare the release:

Release PR:
- [ ] Update Changelog
- [ ] Update README
- [ ] Update Zebra Versions
- [ ] Update End of Support Height

Publish Release:
- [ ] Create & Test GitHub Pre-Release
- [ ] Publish GitHub Release
- [ ] Publish Rust Crates
- [ ] Publish Docker Images

### PR DESCRIPTION
**Pull request by oxarbitrage**
---
name: 'Release Checklist Template'
about: 'Checklist to create and publish a Zebra release'
title: 'Release Zebra (version)'
labels: 'A-release, C-trivial, P-Critical :ambulance:'
assignees: ''

---

# Prepare for the Release

- [x] Make sure there has been [at least one successful full sync test](https://github.com/ZcashFoundation/zebra/actions/workflows/ci-integration-tests-gcp.yml?query=event%3Aschedule) since the last state change, or start a manual full sync.
- [x] Make sure the PRs with the new checkpoint hashes and missed dependencies are already merged.
      (See the release ticket checklist for details)

# Summarise Release Changes

These steps can be done a few days before the release, in the same PR:

## Change Log

**Important**: Any merge into `main` deletes any edits to the draft changelog.
Once you are ready to tag a release, copy the draft changelog into `CHANGELOG.md`.

We use [the Release Drafter workflow](https://github.com/marketplace/actions/release-drafter) to automatically create a [draft changelog](https://github.com/ZcashFoundation/zebra/releases). We follow the [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) format.

To create the final change log:
- [x] Copy the **latest** draft changelog into `CHANGELOG.md` (there can be multiple draft releases)
- [x] Delete any trivial changes
    - [x] Put the list of deleted changelog entries in a PR comment to make reviewing easier
- [x] Combine duplicate changes
- [x] Edit change descriptions so they will make sense to Zebra users
- [x] Check the category for each change
  - Prefer the "Fix" category if you're not sure

## README

README updates can be skipped for urgent releases.

Update the README to:
- [x] Remove any "Known Issues" that have been fixed since the last release.
- [x] Update the "Build and Run Instructions" with any new dependencies.
      Check for changes in the `Dockerfile` since the last tag: `git diff <previous-release-tag> docker/Dockerfile`.
- [x] If Zebra has started using newer Rust language features or standard library APIs, update the known working Rust version in the README, book, and `Cargo.toml`s

You can use a command like:
```sh
fastmod --fixed-strings '1.58' '1.65'
```

## Create the Release PR

- [x] Push the updated changelog and README into a new branch
      for example: `bump-v1.0.0` - this needs to be different to the tag name
- [x] Create a release PR by adding `&template=release-checklist.md` to the comparing url ([Example](https://github.com/ZcashFoundation/zebra/compare/bump-v1.0.0?expand=1&template=release-checklist.md)).
- [x] Freeze the [`batched` queue](https://dashboard.mergify.com/github/ZcashFoundation/repo/zebra/queues) using Mergify.
- [x] Mark all the release PRs as `Critical` priority, so they go in the `urgent` Mergify queue.
- [x] Mark all non-release PRs with `do-not-merge`, because Mergify checks approved PRs against every commit, even when a queue is frozen.

# Update Versions and End of Support

### Choose a Release Level

Zebra follows [semantic versioning](https://semver.org). Semantic versions look like: MAJOR.MINOR.PATCH[-TAG.PRE-RELEASE]

Choose a release level for `zebrad`. Release levels are based on user-visible changes from the changelog:
- Mainnet Network Upgrades are `major` releases
- significant new features or behaviour changes; changes to RPCs, command-line, or configs; and deprecations or removals are `minor` releases
- otherwise, it is a `patch` release

Zebra's Rust API doesn't have any support or stability guarantees, so we keep all the `zebra-*` and `tower-*` crates on a beta `pre-release` version.

### Update Crate Versions

If you're publishing crates for the first time, [log in to crates.io](https://github.com/ZcashFoundation/zebra/blob/doc-crate-own/book/src/dev/crate-owners.md#logging-in-to-cratesio),
and make sure you're a member of owners group.

Check that the release will work:
- [x] Update crate versions, commit the changes to the release branch, and do a release dry-run:

```sh
cargo release version --verbose --execute --allow-branch '*' --workspace --exclude zebrad --exclude zebra-scan beta
# Due to a bug in cargo-release, we need to pass an exact version here
cargo release version --verbose --execute --allow-branch '*' --package zebra-scan 0.1.0-alpha.1
cargo release version --verbose --execute --allow-branch '*' --package zebrad patch # [ major | minor | patch ]
cargo release replace --verbose --execute --allow-branch '*' --package zebrad
cargo release commit --verbose --execute --allow-branch '*'
```

Crate publishing is [automatically checked in CI](https://github.com/ZcashFoundation/zebra/actions/workflows/release-crates-io.yml) using "dry run" mode.

## Update End of Support

The end of support height is calculated from the current blockchain height:
- [x] Find where the Zcash blockchain tip is now by using a [Zcash explorer](https://zcashblockexplorer.com/blocks) or other tool.
- [x] Replace `ESTIMATED_RELEASE_HEIGHT` in [`end_of_support.rs`](https://github.com/ZcashFoundation/zebra/blob/main/zebrad/src/components/sync/end_of_support.rs) with the height you estimate the release will be tagged.

<details>

<summary>Optional: calculate the release tagging height</summary>

- Add `1152` blocks for each day until the release
- For example, if the release is in 3 days, add `1152 * 3` to the current Mainnet block height

</details>

## Update the Release PR

- [x] Push the version increments and the release constants to the release branch.

# Publish the Zebra Release

## Create the GitHub Pre-Release

- [x] Wait for all the release PRs to be merged
- [x] Create a new release using the draft release as a base, by clicking the Edit icon in the [draft release](https://github.com/ZcashFoundation/zebra/releases)
- [x] Set the tag name to the version tag,
      for example: `v1.0.0`
- [x] Set the release to target the `main` branch
- [x] Set the release title to `Zebra ` followed by the version tag,
      for example: `Zebra 1.0.0`
- [x] Replace the prepopulated draft changelog in the release description with the final changelog you created;
      starting just _after_ the title `## [Zebra ...` of the current version being released,
      and ending just _before_ the title of the previous release.
- [x] Mark the release as 'pre-release', until it has been built and tested
- [x] Publish the pre-release to GitHub using "Publish Release"
- [x] Delete all the [draft releases from the list of releases](https://github.com/ZcashFoundation/zebra/releases)

## Test the Pre-Release

- [x] Wait until the Docker binaries have been built on `main`, and the quick tests have passed:
    - [x] [ci-unit-tests-docker.yml](https://github.com/ZcashFoundation/zebra/actions/workflows/ci-unit-tests-docker.yml?query=branch%3Amain)
    - [x] [ci-integration-tests-gcp.yml](https://github.com/ZcashFoundation/zebra/actions/workflows/ci-integration-tests-gcp.yml?query=branch%3Amain)
- [x] Wait until the [pre-release deployment machines have successfully launched](https://github.com/ZcashFoundation/zebra/actions/workflows/cd-deploy-nodes-gcp.yml?query=event%3Arelease)

## Publish Release

- [x] [Publish the release to GitHub](https://github.com/ZcashFoundation/zebra/releases) by disabling 'pre-release', then clicking "Set as the latest release"

## Publish Crates

- [x] [Run `cargo login`](https://github.com/ZcashFoundation/zebra/blob/doc-crate-own/book/src/dev/crate-owners.md#logging-in-to-cratesio)
- [x] Run `cargo clean` in the zebra repo (optional)

The publishing didn't worked for v1.5.1. We made a fix and published v1.5.2, the following steps were done with that tag:

- [x] Publish the crates to crates.io: `cargo release publish --verbose --workspace --execute`
- [x] Check that Zebra can be installed from `crates.io`:
      `cargo install --locked --force --version 1.minor.patch zebrad && ~/.cargo/bin/zebrad`
      and put the output in a comment on the PR.

## Publish Docker Images
- [x] Wait for the [the Docker images to be published successfully](https://github.com/ZcashFoundation/zebra/actions/workflows/release-binaries.yml?query=event%3Arelease).
- [x] Un-freeze the [`batched` queue](https://dashboard.mergify.com/github/ZcashFoundation/repo/zebra/queues) using Mergify.
- [ ] Remove `do-not-merge` from the PRs you added it to

## Release Failures

If building or running fails after tagging:

<details>

<summary>Tag a new release, following these instructions...</summary>

1. Fix the bug that caused the failure
2. Start a new `patch` release
3. Skip the **Release Preparation**, and start at the **Release Changes** step
4. Update `CHANGELOG.md` with details about the fix
5. Follow the release checklist for the new Zebra version

</details>

### CODE REVIEW

**Comment by oxarbitrage** [comment]
Removed from changelog:


- Update copyright year to 2024 ([#8134](https://github.com/ZcashFoundation/zebra/pull/8134))
- build(deps): bump `jsonrpc` from 0.16.0 to 0.17.0 ([#8139](https://github.com/ZcashFoundation/zebra/pull/8139))
- build(deps): bump the devops group with 2 updates ([#8147](https://github.com/ZcashFoundation/zebra/pull/8147))
- build(deps): bump the devops group with 2 updates ([#8126](https://github.com/ZcashFoundation/zebra/pull/8126))
- build(deps): bump the devops group with 2 updates ([#8100](https://github.com/ZcashFoundation/zebra/pull/8100))
- build(deps): bump the devops group with 3 updates ([#8120](https://github.com/ZcashFoundation/zebra/pull/8120))
- build(deps): bump the prod group with 3 updates ([#8158](https://github.com/ZcashFoundation/zebra/pull/8158))
- build(deps): bump the prod group with 4 updates ([#8117](https://github.com/ZcashFoundation/zebra/pull/8117))
- build(deps): bump the prod group with 5 updates ([#8138](https://github.com/ZcashFoundation/zebra/pull/8138))
- build(deps): bump the prod group with 6 updates ([#8125](https://github.com/ZcashFoundation/zebra/pull/8125))
- build(deps): bump zerocopy from 0.7.30 to 0.7.31 ([#8111](https://github.com/ZcashFoundation/zebra/pull/8111))

- cleanup: Fix unused code lints ([#8148](https://github.com/ZcashFoundation/zebra/pull/8148))
- fix(ci): revert script changes causing workflow failures ([#8133](https://github.com/ZcashFoundation/zebra/pull/8133))

- fix(elasticsearch): Make bulk size the same for testnet and mainnet ([#8127](https://github.com/ZcashFoundation/zebra/pull/8127))
- fix(test): Ignore segmentation fault errors when shutting down Docker in tests ([#8107](https://github.com/ZcashFoundation/zebra/pull/8107))
- fix(workflows): mass rename worflows to fix readme badges ([#8116](https://github.com/ZcashFoundation/zebra/pull/8116))
- ref(workflow): move most scripts to their own executables ([#8005](https://github.com/ZcashFoundation/zebra/pull/8005))
- refactor: remove obsolete clippy allows in non\_finalized\_state ([#8140](https://github.com/ZcashFoundation/zebra/pull/8140))
- test(scan): Add raw database format snapshots to the scanner ([#8075](https://github.com/ZcashFoundation/zebra/pull/8075))
- test(scan): Add typed database format snapshots to the scanner ([#8083](https://github.com/ZcashFoundation/zebra/pull/8083))
- test: Add progress markers to the snapshot tests ([#8106](https://github.com/ZcashFoundation/zebra/pull/8106))
- tests(sinsemilla): Minor changes to sinsemilla tests ([#8128](https://github.com/ZcashFoundation/zebra/pull/8128))



### Security

- build(deps): bump zerocopy from 0.7.30 to 0.7.31 ([#8111](https://github.com/ZcashFoundation/zebra/pull/8111))

(Edited to reflect the changes during the review process.)

**Comment by oxarbitrage** [comment]
Temporal failure, retrying: https://github.com/ZcashFoundation/zebra/actions/runs/7574761722/job/20630635394?pr=8165#step:15:96

File: CHANGELOG.md [hunk 1]
```diff
@@ -5,14 +5,38 @@ All notable changes to Zebra are documented in this file.
...
 and this project adheres to [Semantic Versioning](https://semver.org).
 
-## [Zebra 1.6.0](https://github.com/ZcashFoundation/zebra/releases/tag/v1.6.0) - TODO: 2024-01-??
+## [Zebra 1.5.1](https://github.com/ZcashFoundation/zebra/releases/tag/v1.5.1) - 2024-01-18
 
 This release:
-- TODO: summary of other important changes
-- adds an experimental `internal-miner` feature, which mines blocks within `zebrad`. This feature
-  is only supported on testnet. Use a more efficient GPU or ASIC for mainnet mining.
+- Adds a scanner result reader as a zebra utility and finalize MVP scanner features.
```
Comment (upbqdn) [inline CHANGELOG.md]:
```suggestion
- Adds a utility for reading scanning results, and finalizes the MVP features of the scanner.
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -5,14 +5,38 @@ All notable changes to Zebra are documented in this file.
...
+  - Don't scan and log if we are below sapling height ([#8121](https://github.com/ZcashFoundation/zebra/pull/8121))
+  - Restart scanning where left ([#8080](https://github.com/ZcashFoundation/zebra/pull/8080))
+
+### Changed
+
+- Improve feature flag docs ([#8114](https://github.com/ZcashFoundation/zebra/pull/8114))
+- Document how to add a column family ([#8149](https://github.com/ZcashFoundation/zebra/pull/8149))
+- Allow opening the database in a read-only mode ([#8079](https://github.com/ZcashFoundation/zebra/pull/8079))
+- Send all zebrad logs to the journal under systemd ([#7965](https://github.com/ZcashFoundation/zebra/pull/7965))
+- CI workflow: Move most scripts to their own executables ([#8005](https://github.com/ZcashFoundation/zebra/pull/8005))
```
Comment (upbqdn) [inline CHANGELOG.md]: Not user-visible. 

```suggestion
```

**Review by upbqdn — DISMISSED** [review]

**Comment by upbqdn** [comment]
I noticed we mention in the release notes that this release adds the reader of the scanning results, but the PR is excluded from the changelog.

File: CHANGELOG.md [hunk 3]
```diff
@@ -5,14 +5,39 @@ All notable changes to Zebra are documented in this file.
...
+### Added
+
+- Add an internal Zcash miner to Zebra ([#8136](https://github.com/ZcashFoundation/zebra/pull/8136), [#8150](https://github.com/ZcashFoundation/zebra/pull/8150))
+- Blockchain scanner new features:
+  - Don't scan and log if we are below sapling height ([#8121](https://github.com/ZcashFoundation/zebra/pull/8121))
+  - Restart scanning where left ([#8080](https://github.com/ZcashFoundation/zebra/pull/8080))
+  - Add scanning result reader utility ([#8104](https://github.com/ZcashFoundation/zebra/pull/8104), [#8157](https://github.com/ZcashFoundation/zebra/pull/8157))
+
+### Changed
+
```
Comment (upbqdn) [inline CHANGELOG.md:27]:
```suggestion

- Make sure scanner database is accessed using the correct types ([#8112](https://github.com/ZcashFoundation/zebra/pull/8112))
- Move history tree and value balance to typed column families ([#8115](https://github.com/ZcashFoundation/zebra/pull/8115))
- Refactor the user documentation for scanning  ([#8124](https://github.com/ZcashFoundation/zebra/pull/8124))
- Refactor user \& dev documentation ([#8145](https://github.com/ZcashFoundation/zebra/pull/8145))
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -5,14 +5,39 @@ All notable changes to Zebra are documented in this file.
...
+- Adds an experimental `internal-miner` feature, which mines blocks within `zebrad`. This feature is only supported on testnet. Use a more efficient GPU or ASIC for mainnet mining.
+
+### Added
+
+- Add an internal Zcash miner to Zebra ([#8136](https://github.com/ZcashFoundation/zebra/pull/8136), [#8150](https://github.com/ZcashFoundation/zebra/pull/8150))
+- Blockchain scanner new features:
+  - Don't scan and log if we are below sapling height ([#8121](https://github.com/ZcashFoundation/zebra/pull/8121))
+  - Restart scanning where left ([#8080](https://github.com/ZcashFoundation/zebra/pull/8080))
+  - Add scanning result reader utility ([#8104](https://github.com/ZcashFoundation/zebra/pull/8104), [#8157](https://github.com/ZcashFoundation/zebra/pull/8157))
+
```
Comment (upbqdn) [inline CHANGELOG.md:25]:
```suggestion
- Note default path to config in docs ([#8143](https://github.com/ZcashFoundation/zebra/pull/8143))
- Document how to add a column family ([#8149](https://github.com/ZcashFoundation/zebra/pull/8149))

```

File: CHANGELOG.md [hunk 5]
```diff
@@ -5,14 +5,39 @@ All notable changes to Zebra are documented in this file.
...
+- Document how to add a column family ([#8149](https://github.com/ZcashFoundation/zebra/pull/8149))
+- Allow opening the database in a read-only mode ([#8079](https://github.com/ZcashFoundation/zebra/pull/8079))
+- Send all zebrad logs to the journal under systemd ([#7965](https://github.com/ZcashFoundation/zebra/pull/7965))
+
+### Fixed
+
+- Point to a manually created list of Zebra crates in docs ([#8160](https://github.com/ZcashFoundation/zebra/pull/8160))
+- Add shielded-scan.md to the index ([#8095](https://github.com/ZcashFoundation/zebra/pull/8095))
+- Elasticsearch feature, make bulk size the same for testnet and mainnet ([#8127](https://github.com/ZcashFoundation/zebra/pull/8127))
+- Ignore segmentation fault errors when shutting down Docker in tests ([#8107](https://github.com/ZcashFoundation/zebra/pull/8107))
```
Comment (upbqdn) [inline CHANGELOG.md]: Not visible to users.

```suggestion
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -5,14 +5,39 @@ All notable changes to Zebra are documented in this file.
...
+## [Zebra 1.5.1](https://github.com/ZcashFoundation/zebra/releases/tag/v1.5.1) - 2024-01-18
 
 This release:
-- TODO: summary of other important changes
-- adds an experimental `internal-miner` feature, which mines blocks within `zebrad`. This feature
-  is only supported on testnet. Use a more efficient GPU or ASIC for mainnet mining.
 
-TODO: the rest of the changelog
+- Adds a utility for reading scanning results, and finalizes the MVP features of the scanner.
+- Adds an experimental `internal-miner` feature, which mines blocks within `zebrad`. This feature is only supported on testnet. Use a more efficient GPU or ASIC for mainnet mining.
```
Comment (upbqdn) [inline CHANGELOG.md:13]:
```suggestion
- Adds an experimental `internal-miner` feature, which mines blocks within `zebrad`. This feature is only supported on testnet. Use a more efficient GPU or ASIC for mainnet mining.
- Contains many documentation improvements.
```

File: CHANGELOG.md [hunk 7]
```diff
@@ -5,14 +5,39 @@ All notable changes to Zebra are documented in this file.
...
+- Add an internal Zcash miner to Zebra ([#8136](https://github.com/ZcashFoundation/zebra/pull/8136), [#8150](https://github.com/ZcashFoundation/zebra/pull/8150))
+- Blockchain scanner new features:
+  - Don't scan and log if we are below sapling height ([#8121](https://github.com/ZcashFoundation/zebra/pull/8121))
+  - Restart scanning where left ([#8080](https://github.com/ZcashFoundation/zebra/pull/8080))
+  - Add scanning result reader utility ([#8104](https://github.com/ZcashFoundation/zebra/pull/8104), [#8157](https://github.com/ZcashFoundation/zebra/pull/8157))
+
+### Changed
+
+- Improve feature flag docs ([#8114](https://github.com/ZcashFoundation/zebra/pull/8114))
+- Document how to add a column family ([#8149](https://github.com/ZcashFoundation/zebra/pull/8149))
```
Comment (upbqdn) [inline CHANGELOG.md]: Let's move this to the **Added** section.
```suggestion
```

**Review by upbqdn — DISMISSED** [review]
I added a bunch of suggestions.

**Review by upbqdn — APPROVED** [review]

**Comment by upbqdn** [comment]
I edited the lists of PRs not included in the changelog in the comment above to reflect the changes we made in the review.

**Comment by oxarbitrage** [comment]
There is a problem with the publication of the `zebra-chain` crate and i am not sure how to proceed. 

Only the `zebra-test` crate was upgraded by now because the command `cargo release publish --verbose --workspace --execute` fails publishing zebra-chain:

```
  Publishing zebra-chain
    Updating crates.io index
   Packaging zebra-chain v1.0.0-beta.33 (/home/oxarbitrage/zebra/release-1.5.1/main_with_release/zebra/zebra-chain)
   Verifying zebra-chain v1.0.0-beta.33 (/home/oxarbitrage/zebra/release-1.5.1/main_with_release/zebra/zebra-chain)
    Updating crates.io index
error: failed to verify package tarball

Caused by:
  failed to select a version for `equihash`.
      ... required by package `zebra-chain v1.0.0-beta.33 (/home/oxarbitrage/zebra/release-1.5.1/main_with_release/zebra/target/package/zebra-chain-1.0.0-beta.33)`
  versions that meet the requirements `^0.2.0` are: 0.2.0

  the package `zebra-chain` depends on `equihash`, with features: `solver` but `equihash` does not have these features.


  failed to select a version for `equihash` which could resolve this conflict
```

This seems to be a problem introduced with the internal miner. 

I can try to publish the rest of the crates but i think we will have to make a hotfix(1.5.2) anyway.

**Comment by oxarbitrage** [comment]
An easier way could be to use the `--no-verify` flag described in https://crates.io/crates/cargo-release.

If that works we can publish as it is and fix for the next release.

**Pull request merged**